examples/data-objects/datastore-race/src/index.tsx

```diff
@@ -0,0 +1,252 @@
+
+        return dsRuntime;
+    },
+    createLocal: async (containerRuntime: ContainerRuntime) => {
+        const loader = Loader._create(
+            [],
+            [],
+            { load: async () => ({ fluidExport: NormalContainerFactory }) },
+            {},
+            { },
+            new Map(),
+        );
+        const container = await loader.createDetachedContainer(containerRuntime.codeDetails);
+        const res = await container.request({ url: "" });
+        return res.value as MergeableClicker;
+    },
+};
+
+export const NormalContainerFactory: IRuntimeFactory =  {
+    get IRuntimeFactory() {return this;},
+    instantiateRuntime: async (context: IContainerContext)=>{
+        const runtime: ContainerRuntime =  await ContainerRuntime.load(
+            context,
+            [["default", Promise.resolve(MerableClickerInstantiationFactory)]],
+            async (req)=>{
+                const rp = RequestParser.create(req);
+                const ds =  await runtime.getRootDataStore(rp.pathParts[0] ?? "default", true);
+                return ds.request(rp.pathParts.length > 1 ? rp.createSubRequest(1) : { url: "" });
+            },
+            {
+                enableSummarizerNode: false,
+                generateSummaries: false,
+            });
+
+        if (runtime.existing === false) {
+            await runtime.createRootDataStore("default", "default");
+        }
+
+        return runtime;
+    },
+};
+
+export const RaceContainerFactory: IRuntimeFactory = {
+    get IRuntimeFactory() {return this;},
+    instantiateRuntime: async (context: IContainerContext)=>{
+        let current: MergeableClicker;
+        let lastElm: HTMLElement;
+        const viewProxy: IFluidHTMLView = {
+            get IFluidHTMLView() {return this;},
+            render(elm: HTMLElement, options?: IFluidHTMLOptions) {
+                lastElm = elm;
+                current.render(elm);
+            },
+        };
+
+        const runtime =  await ContainerRuntime.load(
+            context,
+            [["default", Promise.resolve(MerableClickerInstantiationFactory)]],
+            async (req)=> {
+                return  { status: 200, mimeType: "fluid/object", value: viewProxy };
+            },
+            {
+                generateSummaries:false,
+                enableSummarizerNode:false,
+            });
+
+        const local = current = await MerableClickerInstantiationFactory.createLocal(runtime);
+
+        runtime.once("connected", ()=>{
+            runtime.once("dataStoreRaceResolved", (id, channel)=>{
+                // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                channel.request({ url:"" }).then(async (req)=>{
+                    const realClicker = req.value as MergeableClicker;
+                    current = realClicker;
+                    ReactDOM.unmountComponentAtNode(lastElm);
+                    local.merge(realClicker);
```

Comment:
I see multiple possible outcomes here:
1. There was no race for first client - its transaction was committed without conflicts. Is this path taken care?
2. There are N clients who lost the race. I assume this is the path they take to merge their state to final object.
3. Client who lost race actually died before seeing race ack. Do we simply drop its state on the floor?

The problem with # 3 is that limited solutions will not scale. The client who is racing need to mostly not care about race and go about its business as if this is already resolved issue. I.e. clients need to be able to store handles to this in-flight data store in other data stores. Maybe it's fine given we are talking about named components (i.e. their handles are the same), but it will be harder for unnamed components (whose handles are stored in map's keys).

For example, at-mentions collection can be created that way, but we should assume that app can put new at-mention instances (guids that are keys into at-mention collection) into Scriptor and those ops should merge just fine. So basically we can't drop # 3 as that requires to also dropping other loosely related ops (addition of at-mentions), we need to ensure that if those ops are already in flight, then some client (ideally all clients) will merge all of these ops correctly.

Also I think for Bohemia the most interesting case is actually not named components - ideally they all should be just created in detached state. It's the **canvas** that is the most interesting and it's "named" due through map's key name, so solution we provide should likely help with that particular flow. If we focus strictly on that, then we already have all the building blocks to implement same solution (CAS via consensus registry plus merge local state in for failed attempt). But we do not have a solution for merge to happen correctly no matter if client died or not.

IMHO, merge should be functional, i.e. not generate ops. be only in-memory process.
That way all clients can do merge and simply rely on eventual consistency.

Reply:
1. In this prototype we always merge as the only real context is the remote context, regardless of which clients op wins the race
2. yes. every client applies their local changes to remote context whenever they get it. These changes propagate normally.
3. in this prototype yes, but the local data store could persist locally. Either like we do for drafts, or potentially like our old detached where we emit ops locally, and those could go to some local place. Could even hook into whatever our iframe data loss story ends up being

yeah. it's tricky. what i like about this is that it is all local which avoids some problems but creates others. I totally agree that anything that can be done while detached should be done while detached. I also prefer functional merge.

For dependent changes i might just block handles to local object from non-local object, but that might quickly get out of control.

i need to think more about non-named creates, but i almost think anything that is shared/coordinated should have a name e.g. you need a name to race

Reply:
a thought. can we just pause outgoing while we are in a race? races should hopefully be rare, and even with other solutions we are essentially blocked until we can round trip. I think this would solve some of the issues above. The next question would then be, can we prepend the merge result, basically rewrite local history before unpausing and sending. this seems fraught with potential issues, but so has every solution to this problem and to me this seems easier to understand/explain

Reply:
I think we need to clearly articulate if this is temp solution (with reasonable next steps / plan) or not.
It may work, but it's a solution for local problem that has global impact.

Better mechanism to pausing is local (to ContainerRuntime) mimicking of dropped connection. That way it has more control on the order of ops on "reconnection" and can stick merged results before the rest of the train.

The part to consider - two such operations in flight - each wants to delay the rest of op processing and stick its own result in front and resume processing. If they are independent, then it's not a problem, but if there is any relationship between them, then they need to continue to be in order relative to each other.

Another thing to consider - it does not compose with batches - these ops can't be done in the batch.

Reply:
yes. i was using pause as shorthand, specifically return false for canSendOps while there is an active race, so we can batch at the container runtime level. 

I need to think more about parallel merges. My general thought is that merges are functional, but output ops (the diff). Merges should also be synchronous. The merge ops are submitted before outstanding ops. For outstanding ops the original batching should be preserved. The race would also have to be submitted at a batch boundary, so we don't split a batch, which could also have interesting implications

Reply:
In the above we would also need to regenerate all pending ops after the merge before sending them, so they appropriately take into account the merge diffs